### PR TITLE
sql: add support for LIMIT with DELETE

### LINKED ISSUE #18006: sql: support LIMIT for DELETEs
MySQL has a `LIMIT` clause for `DELETE` that allows one to incrementally delete rows from a table. Instead of deleting all rows at once, one can execute `DELETE ... LIMIT N` until there are no more rows to be deleted. This reduces the duration of the transaction containing the `DELETE`. (See the [MySQL DELETE docs](https://dev.mysql.com/doc/refman/5.7/en/delete.html) for more details.)

This is a workaround for large `DELETE`s until we have support for [support for larger transactions](https://github.com/cockroachdb/cockroach/issues/15849). This would improve the user experience described in #17921.

The one downside I can think of this approach is that [Postgres's DELETE](https://www.postgresql.org/docs/9.6/static/sql-delete.html) does not support the `LIMIT` clause, which means that ORMs' Postgres dialects won't generate queries that use `LIMIT`s.

### PR DESCRIPTION
**Pull request by knz**
Suggested by @cuongdo:

> MySQL has a LIMIT clause for DELETE that allows one to incrementally
> delete rows from a table. Instead of deleting all rows at once, one
> can execute DELETE ... LIMIT N until there are no more rows to be
> deleted. This reduces the duration of the transaction containing the
> DELETE.

Fixes #18006.
Alleviates #17921.

### CODE REVIEW

**Comment by tschottdorf** [comment]
Very happy you put this together so quickly! Definitely has my blessing.

File: pkg/sql/logictest/testdata/logic_test/delete [hunk 1]
```diff
@@ -173,3 +173,54 @@ CREATE TABLE indexed (id int primary key, value int, other int, index (value))
...
+0  ·       from   indexed
+1  limit   ·      ·
+2  render  ·      ·
+3  scan    ·      ·
+3  ·       table  indexed@primary
+3  ·       spans  ALL
+3  ·       limit  10
+
+query ITTT
+EXPLAIN DELETE FROM indexed WHERE value = 5 LIMIT 10 RETURNING id
```
Comment (justinj) [inline pkg/sql/logictest/testdata/logic_test/delete]: Could you also put in a test of actually executing `DELETE...LIMIT`? If the order being undefined means you can't check the exact results you can check the `COUNT(*)`, right?

File: pkg/sql/parser/sql.y [hunk 2]
```diff
@@ -1425,12 +1425,19 @@ create_stmt:
...
+//               [LIMIT <expr>]
+//               [RETURNING <exprs...>]
 // %SeeAlso: WEBDOCS/delete.html
 delete_stmt:
-  opt_with_clause DELETE FROM relation_expr_opt_alias where_clause returning_clause
+  opt_with_clause DELETE FROM relation_expr_opt_alias where_clause opt_limit_clause returning_clause
   {
-    $$.val = &Delete{Table: $4.tblExpr(), Where: newWhere(astWhere, $5.expr()), Returning: $6.retClause()}
+    $$.val = &Delete{
+               Table: $4.tblExpr(),
```
Comment (justinj) [inline pkg/sql/parser/sql.y]: nit: I think the indentation here is a little wonky.

**Comment by justinj** [comment]
Hah, what a pleasantly small change!

**Comment by knz** [comment]
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/delete, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18023#-KsnKbrePCfihPE3jMQM-r1-215:-KsnTuA2KUfAQoD4apLi:b-8vskqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b221ad0524cdb33b97cc9a3b922f3ee1deeb1e76/pkg/sql/logictest/testdata/logic_test/delete#L215)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Could you also put in a test of actually executing `DELETE...LIMIT`? If the order being undefined means you can't check the exact results you can check the `COUNT(*)`, right?
</blockquote></details>

The result tag contains the number of rows affected, or, if RETURNING was used, the rows actually deleted. The client has plenty of information to work with. Unfortunately, we can't check the tag in logic tests, but I put a RETURNING test here to showcase. Also indeed, the rows are non-deterministic so we need count(*).

---

*[pkg/sql/parser/sql.y, line 1436 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18023#-KsnKbrePCfihPE3jMQR-r1-1436:-KsnTsfJQRtMAFfAkRZb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b221ad0524cdb33b97cc9a3b922f3ee1deeb1e76/pkg/sql/parser/sql.y#L1436)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: I think the indentation here is a little wonky.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18023)*

**Comment by justinj** [comment]
LGTM!

**Comment by knz** [comment]
TFYR!

**Comment by bdarnell** [comment]
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/delete, line 215 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18023#-KsnKbrePCfihPE3jMQM-r1-215:-Kso0Df4xXPm-vLKBRTX:baheigp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b221ad0524cdb33b97cc9a3b922f3ee1deeb1e76/pkg/sql/logictest/testdata/logic_test/delete#L215)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The result tag contains the number of rows affected, or, if RETURNING was used, the rows actually deleted. The client has plenty of information to work with. Unfortunately, we can't check the tag in logic tests, but I put a RETURNING test here to showcase. Also indeed, the rows are non-deterministic so we need count(*).
</blockquote></details>

It's probably not worth supporting here, but mysql also extends DELETE with an ORDER BY clause to make the results of a LIMITed DELETE deterministic. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18023)*

**Pull request merged**